src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -1149,23 +1149,47 @@ public override BoundNode VisitDynamicInvocation(BoundDynamicInvocation node)
             return null;
         }
 
-        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node)
+        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node, InterpolatedStringHandlerData? nullableData)
         {
-            foreach (var expr in node.Parts)
+            // If there can be any branching, then we need to treat the expressions
+            // as optionally evaluated. Otherwise, we treat them as always evaluated
+            switch (nullableData)
             {
-                VisitRvalue(expr);
+                case null:
+                    visitParts();
+                    break;
+                case { HasTrailingHandlerValidityParameter: false, UsesBoolReturns: false, Construction: var construction }:
+                    VisitRvalue(construction);
+                    visitParts();
+                    break;
+                case { Construction: var construction }:
+                    VisitRvalue(construction);
+                    var beforePartsState = State.Clone();
+                    visitParts();
```

Comment:
Feels like we should join the state before and after each part in this case when `UsesBoolReturns` is true. I think this would affect nullability analysis and maybe also region analysis.
Also, if I understood coorectly, the handler constructor having a success value is independant of the append calls having a success value. We probably need to handle those permutations.

Reply:
I don't think this can actually have an effect.

Reply:
I think it should be possible to observe. Try putting assignments in the holes (`$"{x = null} {x = string.Empty}"`) and observe `x` afterwards.